File: mobile/ios/Sources/MobileLocalization.swift
import Foundation
import SwiftUI

/// The mobile language list mirrors macOS AppLanguage.  `system` is a
/// selection mode; the other 30 entries are concrete resource locales.
enum MobileLanguage: String, CaseIterable, Identifiable, Codable, Sendable {
    case system
    case tr, en, de, es, fr, it, ja, ko, nl, pt, ru
    case zhHans = "zh-Hans"
    case ar, bn, hi, id, vi, ur, mr, te, ta, fa, pl, uk, th, ms, ro, el, cs, hu

    var id: String { rawValue }

    static var supported: [MobileLanguage] {
        allCases.filter { $0 != .system }
    }

    var localeIdentifier: String {
        switch self {
        case .system: return MobileLanguage.from(preferredIdentifier: Locale.preferredLanguages.first ?? "en").localeIdentifier
        case .tr: return "tr-TR"
        case .en: return "en-US"
        case .de: return "de-DE"
        case .es: return "es-ES"
        case .fr: return "fr-FR"
        case .it: return "it-IT"
        case .ja: return "ja-JP"
        case .ko: return "ko-KR"
        case .nl: return "nl-NL"
        case .pt: return "pt-PT"
        case .ru: return "ru-RU"
        case .zhHans: return "zh-CN"
        case .ar: return "ar-SA"
        case .bn: return "bn-BD"
        case .hi: return "hi-IN"
        case .id: return "id-ID"
        case .vi: return "vi-VN"
        case .ur: return "ur-PK"
        case .mr: return "mr-IN"
        case .te: return "te-IN"
        case .ta: return "ta-IN"
        case .fa: return "fa-IR"
        case .pl: return "pl-PL"
        case .uk: return "uk-UA"
        case .th: return "th-TH"
        case .ms: return "ms-MY"
        case .ro: return "ro-RO"
        case .el: return "el-GR"
        case .cs: return "cs-CZ"
        case .hu: return "hu-HU"
        }
    }

    var nativeName: String {
        switch self {
        case .system: return "System language"
        case .tr: return "Türkçe"
        case .en: return "English"
        case .de: return "Deutsch"
        case .es: return "Español"
        case .fr: return "Français"
        case .it: return "Italiano"
        case .ja: return "日本語"
        case .ko: return "한국어"
        case .nl: return "Nederlands"
        case .pt: return "Português"
        case .ru: return "Русский"
        case .zhHans: return "简体中文"
        case .ar: return "العربية"
        case .bn: return "বাংলা"
        case .hi: return "हिन्दी"
        case .id: return "Bahasa Indonesia"
        case .vi: return "Tiếng Việt"
        case .ur: return "اردو"
        case .mr: return "मराठी"
        case .te: return "తెలుగు"
        case .ta: return "தமிழ்"
        case .fa: return "فارسی"
        case .pl: return "Polski"
        case .uk: return "Українська"
        case .th: return "ไทย"
        case .ms: return "Bahasa Melayu"
        case .ro: return "Română"
        case .el: return "Ελληνικά"
        case .cs: return "Čeština"
        case .hu: return "Magyar"
        }
    }

    var flagEmoji: String {
        switch self {
        case .system: return "🌐"
        case .tr: return "🇹🇷"
        case .en: return "🇺🇸"
        case .de: return "🇩🇪"
        case .es: return "🇪🇸"
        case .fr: return "🇫🇷"
        case .it: return "🇮🇹"
        case .ja: return "🇯🇵"
        case .ko: return "🇰🇷"
        case .nl: return "🇳🇱"
        case .pt: return "🇵🇹"
        case .ru: return "🇷🇺"
        case .zhHans: return "🇨🇳"
        case .ar: return "🇸🇦"
        case .bn: return "🇧🇩"
        case .hi, .mr, .te, .ta: return "🇮🇳"
        case .id: return "🇮🇩"
        case .vi: return "🇻🇳"
        case .ur: return "🇵🇰"
        case .fa: return "🇮🇷"
        case .pl: return "🇵🇱"
        case .uk: return "🇺🇦"
        case .th: return "🇹🇭"
        case .ms: return "🇲🇾"
        case .ro: return "🇷🇴"
        case .el: return "🇬🇷"
        case .cs: return "🇨🇿"
        case .hu: return "🇭🇺"
        }
    }

    var isRTL: Bool {
        switch self {
        case .ar, .ur, .fa: return true
        case .system: return MobileLanguage.from(preferredIdentifier: Locale.preferredLanguages.first ?? "en").isRTL
        default: return false
        }
    }

    var shortCode: String {
        switch self {
        case .system: return "AUTO"
        case .zhHans: return "ZH"
        default: return rawValue.uppercased()
        }
    }

    static func from(preferredIdentifier identifier: String) -> MobileLanguage {
        let normalized = identifier.replacingOccurrences(of: "_", with: "-").lowercased()
        if normalized.hasPrefix("zh-") || normalized == "zh" { return .zhHans }
        return supported.first {
            let code = $0.rawValue.lowercased()
            return normalized == code || normalized.hasPrefix(code + "-")
        } ?? .en
    }

    static func stored(_ rawValue: String?) -> MobileLanguage {
        guard let rawValue, let value = MobileLanguage(rawValue: rawValue) else { return .system }
        return value
    }
}

@MainActor
final class MobileLocalization: ObservableObject {
    static let storageKey = "mobile.language"

    @Published private(set) var selected: MobileLanguage {
        didSet { defaults.set(selected.rawValue, forKey: Self.storageKey) }
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        selected = MobileLanguage.stored(defaults.string(forKey: Self.storageKey))
    }

    var effective: MobileLanguage {
        selected == .system
            ? MobileLanguage.from(preferredIdentifier: Locale.preferredLanguages.first ?? "en")
            : selected
    }

    var locale: Locale { Locale(identifier: effective.localeIdentifier) }
    var isRTL: Bool { effective.isRTL }

    func set(_ language: MobileLanguage) {
        selected = language
    }

    func text(_ key: String) -> String {
        MobileCopy.text(key, language: effective)
    }

    func format(_ key: String, _ arguments: CVarArg...) -> String {
        MobileCopy.format(key, language: effective, locale: locale, arguments: arguments)
    }
}

enum MobileCopy {
    static func bundle(for language: MobileLanguage) -> Bundle {
        let path = Bundle.main.path(forResource: language.rawValue, ofType: "lproj")
            ?? Bundle.main.path(forResource: language.rawValue.lowercased(), ofType: "lproj")
        return path.flatMap(Bundle.init(path:)) ?? .main
    }

    static func text(_ key: String, language: MobileLanguage = .en) -> String {
        let localized = bundle(for: language).localizedString(forKey: key, value: nil, table: nil)
        if localized != key { return localized }
        return Bundle.main.localizedString(forKey: key, value: key, table: nil)
    }

    static func format(
        _ key: String,
        language: MobileLanguage = .en,
        locale: Locale = .current,
        arguments: [CVarArg]
    ) -> String {
        String(format: text(key, language: language), locale: locale, arguments: arguments)
    }
}

enum MobileIntroNavigation {
    static func canSkip(page: Int, pageCount: Int) -> Bool { page < pageCount - 1 }
    static func showsBack(page: Int) -> Bool { page > 0 }
    static func primaryKey(page: Int, pageCount: Int) -> String {
        page == pageCount - 1 ? "intro.start" : "intro.continue"
    }
}

struct MobileLanguagePicker: View {
    @EnvironmentObject private var localization: MobileLocalization
    @State private var isPresented = false
    var darkAppearance = false

    var body: some View {
        Button {
            isPresented = true
        } label: {
            Label(localization.selected.shortCode, systemImage: "globe")
                .font(.system(size: 13, weight: .semibold, design: .rounded))
                .labelStyle(.titleAndIcon)
                .foregroundStyle(darkAppearance ? .white : .primary)
                .padding(.horizontal, 10)
                .padding(.vertical, 7)
                .background((darkAppearance ? Color.white.opacity(0.16) : Color.primary.opacity(0.08)))
                .clipShape(Capsule())
        }
        .buttonStyle(.plain)
        .accessibilityLabel(localization.text("intro.languageAccessibility"))
        .sheet(isPresented: $isPresented) {
            NavigationStack {
                List {
                    languageRow(.system)
                    ForEach(MobileLanguage.supported) { languageRow($0) }
                }
                .navigationTitle(localization.text("intro.language"))
                .navigationBarTitleDisplayMode(.inline)
            }
            .presentationDetents([.medium, .large])
        }
    }

    @ViewBuilder
    private func languageRow(_ language: MobileLanguage) -> some View {
        Button {
            localization.set(language)
            isPresented = false
        } label: {
            HStack(spacing: 12) {
                Text(language.flagEmoji)
                Text(language == .system ? localization.text("intro.languageSystem") : language.nativeName)
                Spacer()
                if localization.selected == language {
                    Image(systemName: "checkmark")
                        .foregroundStyle(.tint)
                }
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

File: mobile/android/app/src/main/java/com/dots/herness/mobile/MobileLocalization.kt
package com.dots.herness.mobile

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import android.os.LocaleList
import java.util.Locale

data class MobileLanguage(
    val code: String,
    val localeTag: String,
    val nativeName: String,
    val flag: String,
    val isRtl: Boolean,
) {
    val shortCode: String get() = if (code == "system") "AUTO" else if (code == "zh-Hans") "ZH" else code.uppercase(Locale.ROOT)
}

object MobileLanguages {
    val system = MobileLanguage("system", "system", "System language", "🌐", false)

    // Keep this order identical to macOS AppLanguage and the shared JSON manifest.
    val supported = listOf(
        MobileLanguage("tr", "tr-TR", "Türkçe", "🇹🇷", false),
        MobileLanguage("en", "en-US", "English", "🇺🇸", false),
        MobileLanguage("de", "de-DE", "Deutsch", "🇩🇪", false),
        MobileLanguage("es", "es-ES", "Español", "🇪🇸", false),
        MobileLanguage("fr", "fr-FR", "Français", "🇫🇷", false),
        MobileLanguage("it", "it-IT", "Italiano", "🇮🇹", false),
        MobileLanguage("ja", "ja-JP", "日本語", "🇯🇵", false),
        MobileLanguage("ko", "ko-KR", "한국어", "🇰🇷", false),
        MobileLanguage("nl", "nl-NL", "Nederlands", "🇳🇱", false),
        MobileLanguage("pt", "pt-PT", "Português", "🇵🇹", false),
        MobileLanguage("ru", "ru-RU", "Русский", "🇷🇺", false),
        MobileLanguage("zh-Hans", "zh-CN", "简体中文", "🇨🇳", false),
        MobileLanguage("ar", "ar-SA", "العربية", "🇸🇦", true),
        MobileLanguage("bn", "bn-BD", "বাংলা", "🇧🇩", false),
        MobileLanguage("hi", "hi-IN", "हिन्दी", "🇮🇳", false),
        MobileLanguage("id", "id-ID", "Bahasa Indonesia", "🇮🇩", false),
        MobileLanguage("vi", "vi-VN", "Tiếng Việt", "🇻🇳", false),
        MobileLanguage("ur", "ur-PK", "اردو", "🇵🇰", true),
        MobileLanguage("mr", "mr-IN", "मराठी", "🇮🇳", false),
        MobileLanguage("te", "te-IN", "తెలుగు", "🇮🇳", false),
        MobileLanguage("ta", "ta-IN", "தமிழ்", "🇮🇳", false),
        MobileLanguage("fa", "fa-IR", "فارسی", "🇮🇷", true),
        MobileLanguage("pl", "pl-PL", "Polski", "🇵🇱", false),
        MobileLanguage("uk", "uk-UA", "Українська", "🇺🇦", false),
        MobileLanguage("th", "th-TH", "ไทย", "🇹🇭", false),
        MobileLanguage("ms", "ms-MY", "Bahasa Melayu", "🇲🇾", false),
        MobileLanguage("ro", "ro-RO", "Română", "🇷🇴", false),
        MobileLanguage("el", "el-GR", "Ελληνικά", "🇬🇷", false),
        MobileLanguage("cs", "cs-CZ", "Čeština", "🇨🇿", false),
        MobileLanguage("hu", "hu-HU", "Magyar", "🇭🇺", false),
    )

    val all: List<MobileLanguage> = listOf(system) + supported

    fun fromCode(code: String?): MobileLanguage = all.firstOrNull { it.code == code } ?: system

    fun fromDevice(locales: LocaleList = LocaleList.getDefault()): MobileLanguage {
        return fromDevice((0 until locales.size()).map { locales[it] })
    }

    fun fromDevice(locales: List<Locale>): MobileLanguage {
        for (locale in locales) {
            val language = locale.toLanguageTag().replace('_', '-').lowercase(Locale.ROOT)
            if (language == "zh" || language.startsWith("zh-")) return fromCode("zh-Hans")
            supported.firstOrNull { code -> language == code.code.lowercase(Locale.ROOT) || language.startsWith(code.code.lowercase(Locale.ROOT) + "-") }
                ?.let { return it }
        }
        return fromCode("en")
    }

    fun effective(selected: MobileLanguage, locales: LocaleList = LocaleList.getDefault()): MobileLanguage =
        if (selected.code == system.code) fromDevice(locales) else selected
}

object MobileIntroNavigation {
    fun canSkip(page: Int, pageCount: Int): Boolean = page < pageCount - 1
    fun showsBack(page: Int): Boolean = page > 0
    fun primaryKey(page: Int, pageCount: Int): String = if (page == pageCount - 1) "intro_start" else "intro_continue"
}

fun SharedPreferences.mobileLanguage(): MobileLanguage = MobileLanguages.fromCode(getString(MOBILE_LANGUAGE_KEY, "system"))

fun SharedPreferences.setMobileLanguage(language: MobileLanguage) {
    edit().putString(MOBILE_LANGUAGE_KEY, language.code).apply()
}

/**
 * Android resource lookup is tied to a Context.  A copied Configuration keeps
 * the app's current resources intact while Compose renders through the
 * selected locale context.
 */
fun Context.createMobileLocaleContext(language: MobileLanguage): Context {
    if (language.code == "system") return this
    val configuration = Configuration(resources.configuration)
    configuration.setLocale(Locale.forLanguageTag(language.localeTag))
    configuration.setLayoutDirection(Locale.forLanguageTag(language.localeTag))
    return createConfigurationContext(configuration)
}
